### PR TITLE
Implement `distinctBy` for SeqLike and Stream

### PR DESCRIPTION
**Pull request by vergenzt**
Opening against 2.12.x as requested from #4612.

> Hi all! I wasn't sure what the right order to do things in for a PR would be, but the readme seemed to indicate it'd be fine to open a PR first and figure things out later. :)
> - I have not yet implemented tests, but I wasn't sure where those should go. Any guidance on that?
> - ~~"The current maintenance release (2.11.x) cannot break source/binary compatibility, which means public APIs cannot change." Does adding a method break compatibility?~~
> 
> @Ichoran and @axel22, the readme and PR policy doc mentioned you're the people to ping on the collections library (I think). Any thoughts?
> 
> Thanks for reading!

### CODE REVIEW

**Comment by vergenzt** [comment]
Should I open a Jira issue/start a scala-internals thread for something like this?

**Comment by axel22** [comment]
Please add some unit tests to test the new functionality, both on general sequences and on streams. Add the tests to `test/junit/scala/collection`.

**Comment by axel22** [comment]
Please do a one-off experiment to assess the performance of your change - you should compare the previous `distinct` operation running time with the new implementation.

File: src/library/scala/collection/immutable/Stream.scala [hunk 1]
```diff
@@ -933,13 +933,34 @@ self =>
    * // produces: "1, 2, 3, 4, 5, 6"
    * }}}
    */
-  override def distinct: Stream[A] = {
+  override def distinct: Stream[A] = distinctBy(identity)
+
+  /** Builds a new stream from this stream in which elements with duplicate
+   * outputs by `f` are removed. Among elements with the same result by `f`,
```
Comment (axel22) [inline src/library/scala/collection/immutable/Stream.scala:939]: Please indent this text to the same indentation level as the first line.

File: src/library/scala/collection/SeqLike.scala [hunk 2]
```diff
@@ -499,13 +499,27 @@ trait SeqLike[+A, +Repr] extends Any with IterableLike[A, Repr] with GenSeqLike[
...
+
+  /** Builds a new $coll from this $coll where output elements have distinct
+   *  results by `f`.
+   *  $willNotTerminateInf
+   *
+   *  @return  A new $coll which contains the first occurence of each element
+   *    `x` with a unique `y = f(x)`.
+   *  @example {{{
+   *    List("Steve", "Barbara", "Stephanie", "Bob", "Frank").distinctBy(_.head) =
+   *    List("Steve", "Barbara", "Frank")
```
Comment (axel22) [inline src/library/scala/collection/SeqLike.scala:512]: Please fix indentation here.

File: src/library/scala/collection/immutable/Stream.scala [hunk 3]
```diff
@@ -933,13 +933,34 @@ self =>
...
+   * }}}
+   */
+  override def distinctBy[B](f: A => B): Stream[A] = {
     // This should use max memory proportional to N, whereas
     // recursively calling distinct on the tail is N^2.
-    def loop(seen: Set[A], rest: Stream[A]): Stream[A] = {
-      if (rest.isEmpty) rest
-      else if (seen(rest.head)) loop(seen, rest.tail)
-      else cons(rest.head, loop(seen + rest.head, rest.tail))
+    def loop(seen: Set[B], rest: Stream[A]): Stream[A] = rest match {
```
Comment (axel22) [inline src/library/scala/collection/immutable/Stream.scala:956]: Is there any strong reason to change the previous implementation to the one with pattern matching?
Please use the previous recursive implementation if not.

File: src/library/scala/collection/immutable/Stream.scala [hunk 4]
```diff
@@ -933,13 +933,34 @@ self =>
    * // produces: "1, 2, 3, 4, 5, 6"
    * }}}
    */
-  override def distinct: Stream[A] = {
+  override def distinct: Stream[A] = distinctBy(identity)
+
+  /** Builds a new stream from this stream in which elements with duplicate
```
Comment (axel22) [inline src/library/scala/collection/immutable/Stream.scala:938]: The first line always shows in ScalaDoc, and is meant to be a short self-contained explanation of the functionality. Please shorten the line to something like this:

```
Builds a new stream in which elements mapped to the same value with function `f` are de-duplicated.
```

The, separate paragraph can describe semantics in more detail, e.g. the first element is always kept, etc.

**Comment by axel22** [comment]
Thanks for your contribution, nice work!

We would need to add unit tests for this, and a one-off benchmark, to ensure that `distinct` didn't change performance significantly, would be helpful.

About scala-internals - yes, generally changes in the collections API should be first discussed on scala-internals, and then a Jira ticket opened.

**Comment by vergenzt** [comment]
Thanks for the feedback! A few thoughts/questions:
- Should I just create a new JUnit test file? There don't seem to be tests for any related methods there. (Notably, `distinct` itself is not tested there. The only place I see `distinct` tested is in [`test/files/run/colltest1.scala`](https://github.com/scala/scala/blob/2.11.x/test/files/run/colltest1.scala). Would it not be better to add tests there?)
- I'll start a thread on scala-internals before doing too much more.
  
  Thanks again!

**Comment by axel22** [comment]
Yes, `colltest1` is a better idea!

Thanks for discussing on scala-internals!

**Comment by Ichoran** [comment]
Incidentally, if doing the benchmarks is going to be too arduous, I can do it when (if?) I review the PR.

**Comment by vergenzt** [comment]
@Ichoran, that would be lovely! Thanks! I'd be happy to do it myself but I imagine it would take much longer. :-\

**Comment by jozic** [comment]
there is more to add/change (unless something drastically changed since last time we tried to push distinctBy in collections :))
see here https://github.com/scala/scala/pull/3850

**Comment by vergenzt** [comment]
Ahh thanks for that reference. I had tried to search for discussions and previous attempts didn't find that, nor the mailing list discussion.

That looks like a much more thorough  implementation, so if it's decided that this is a good idea to merge it'd probably make more  sense to merge that implementation. We can close this.

**Pull request closed without merging**